core/src/test/scala/integration/kafka/server/FetchFromFollowerIntegrationTest.scala

```diff
@@ -0,0 +1,106 @@
+  val followerBrokerId = 1
+  var admin: Admin = null
+
+  def overridingProps: Properties = {
+    val props = new Properties
+    props.put(KafkaConfig.NumPartitionsProp, numParts.toString)
+
+    props
+  }
+
+  override def generateConfigs: collection.Seq[KafkaConfig] =
+    TestUtils.createBrokerConfigs(numNodes, zkConnectOrNull, enableControlledShutdown = false, enableFetchFromFollower = true)
+      .map(KafkaConfig.fromProps(_, overridingProps))
+
+  // Create a 2 broker cluster where broker 0 is the leader and 1 is the follower.
+  // Create a topic with a single partition. Access brokers via `brokers`
+  @BeforeEach
+  def initializeFetchFromFollowerCluster(): Unit = {
+    createTopic(topic, numParts, numNodes)
+    TestUtils.generateAndProduceMessages(brokers, topic, initialMessages)
+
+    // Make broker 1 the follower
+    admin = TestUtils.createAdminClient(brokers, listenerName)
+    val topicDescription = TestUtils.describeTopic(admin, topic)
+    topicDescription.partitions.forEach(partition => {
+      // only reelect if leader is not broker 0
+      if (partition.leader().id() != leaderBrokerId) {
+        val reassignment = Map(new TopicPartition(topic, 0) -> Optional.of(
+          new NewPartitionReassignment(List[Integer](leaderBrokerId, followerBrokerId).asJava))).asJava
+        admin.alterPartitionReassignments(reassignment).all.get
+        TestUtils.waitUntilTrue(() => {
+          val reassigned = admin.listPartitionReassignments.reassignments.get
+          reassigned.isEmpty
+        }, "reassignment did not complete.")
+        admin.electLeaders(ElectionType.PREFERRED, Set(new TopicPartition(topic, 0)).asJava).all.get
+      }
+    })
+
+    // Check the log size for each broker so that we can distinguish between failures caused by replication issues
+    // versus failures caused by the metrics
+    val topicPartition = new TopicPartition(topic, 0)
+    brokers.foreach { broker =>
+      val log = broker.logManager.getLog(topicPartition)
+      val brokerId = broker.config.brokerId
+      val logSize = log.map(_.size)
+      assertTrue(logSize.exists(_ > 0), s"Expected broker $brokerId to have a Log for $topicPartition with positive size, actual: $logSize")
+    }
+  }
+
+  @AfterEach
+  def close(): Unit = {
+    if (admin != null) {
+      admin.close()
+    }
+  }
+
+  @ParameterizedTest(name = TestInfoUtils.TestWithParameterizedQuorumName)
+  @ValueSource(strings = Array("zk", "kraft"))
+  def testFollowerCompleteDelayedPurgatoryOnReplication(quorum: String): Unit = {
+    TestUtils.generateAndProduceMessages(brokers, topic, nMessages)
+    // set fetch.max.wait.ms to a value greater than test utils wait time (15 seconds) to ensure that the
+    // delayed fetch purgatory is completed after successfully replicating from the leader.
+    val messages = initialMessages + nMessages
+    TestUtils.consumeTopicRecords(brokers, topic, messages, rackId = followerBrokerId.toString, fetchMaxWaitMs = 60000)
```

Comment:
Did you consider sending a FetchRequest directly instead of relying on the consumer here? The downside of the using the consumer is that it does really guarantee that the fetch is serviced from the follower. It could be serviced by the leader if the cluster is misconfigured for instance.

Reply:
unrelated question: i was looking into verifying the follower's bytes out metric but was a bit confused. in https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/metrics/MetricsTest.scala#L193 we set up a 2 broker cluster and access the bytes out metric using `TestUtils.meterCount()` which accesses the static yammer metrics registry. 

for integration tests, does this mean brokers' yammer metrics are merged together into one registry?